File: box/src/main/java/de/qabel/box/storage/DeleteOnCloseFileInputStream.java
package de.qabel.box.storage;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;

public class DeleteOnCloseFileInputStream extends FileInputStream {
    private static final Logger logger = LoggerFactory.getLogger(DeleteOnCloseFileInputStream.class);
    private final File file;

    public DeleteOnCloseFileInputStream(String name) throws FileNotFoundException {
        this(new File(name));
    }

    public DeleteOnCloseFileInputStream(File file) throws FileNotFoundException {
        super(file);
        this.file = file;
    }

    @Override
    public void close() throws IOException {
        try {
            super.close();
        } finally {
            try {
                file.delete();
            } catch (Exception e) {
                logger.warn(
                    "failed to delete tmp file for download: " + file.getAbsolutePath() + ": " + e.getMessage(),
                    e
                );
            }
        }
    }
}

File: box/src/main/java/de/qabel/box/storage/DeleteOnCloseFileInputStream.kt
package de.qabel.box.storage

import org.slf4j.Logger
import org.slf4j.LoggerFactory

import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException

class DeleteOnCloseFileInputStream @Throws(FileNotFoundException::class)
constructor(private val file: File) : FileInputStream(file) {

    @Throws(FileNotFoundException::class)
    constructor(name: String) : this(File(name)) {
    }

    @Throws(IOException::class)
    override fun close() {
        try {
            super.close()
        } finally {
            try {
                file.delete()
            } catch (e: Exception) {
                logger.warn(
                        "failed to delete tmp file for download: " + file.absolutePath + ": " + e.message,
                        e)
            }

        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DeleteOnCloseFileInputStream::class.java)
    }
}
